projects/packages/backup-helper-script-manager/src/class-helper-script-manager.php

```diff
@@ -0,0 +1,584 @@
+					$wp_content_url = $abspath_url . $wp_content_url;
+				}
+
+				$locations[ $wp_content_dir ] = $wp_content_url;
+			}
+		} catch ( Exception $exception ) {
+			$locations[ \WP_CONTENT_DIR ] = new \WP_Error(
+				'content_path_missing',
+				'Unable to access content path "' . \WP_CONTENT_DIR . '"' . $exception->getMessage(),
+				array( 'status' => 500 )
+			);
+		}
+
+		$upload_dir_info = \wp_upload_dir();
+		$wp_uploads_dir  = $upload_dir_info['basedir'];
+
+		try {
+			if ( Throw_On_Errors::t_is_dir( $wp_uploads_dir ) ) {
+
+				$wp_uploads_dir = Throw_On_Errors::t_realpath( $wp_uploads_dir );
+				$wp_uploads_url = $upload_dir_info['baseurl'];
+
+				if ( ! wp_http_validate_url( $wp_uploads_url ) ) {
+					$wp_uploads_url = $abspath_url . $wp_uploads_url;
+				}
+
+				$locations[ $wp_uploads_dir ] = $wp_uploads_url;
+			}
+		} catch ( Exception $exception ) {
+			$locations[ $wp_uploads_dir ] = new \WP_Error(
+				'uploads_path_missing',
+				'Unable to access uploads path "' . $wp_uploads_dir . '"' . $exception->getMessage(),
+				array( 'status' => 500 )
+			);
+		}
+
+		try {
+			if ( Throw_On_Errors::t_is_dir( \ABSPATH ) ) {
+				$abspath_dir               = Throw_On_Errors::t_realpath( \ABSPATH );
+				$locations[ $abspath_dir ] = $abspath_url;
+			}
+		} catch ( Exception $exception ) {
+			$locations[ \ABSPATH ] = new \WP_Error(
+				'abspath_missing',
+				'Unable to access WordPress root "' . \ABSPATH . '": ' . $exception->getMessage(),
+				array( 'status' => 500 )
+			);
+		}
+
+		return $locations;
+	}
+
+	/**
+	 * Installs a Helper Script, and returns its filesystem path and access url.
+	 *
+	 * @param string $script_body Helper Script file contents.
+	 *
+	 * @return array|\WP_Error Either an array containing the filesystem path ("path"), the URL ("url") of the helper
+	 *   script, and the WordPress root ("abspath"), or an instance of WP_Error.
+	 */
+	public function install_helper_script( $script_body ) {
+		// Check that the script body contains the correct header.
+		$actual_header = static::string_starts_with_substring( $script_body, static::HELPER_HEADER );
+		if ( true !== $actual_header ) {
+			return new \WP_Error(
+				'bad_header',
+				'Bad helper script header: 0x' . bin2hex( $actual_header ),
+				array( 'status' => 400 )
+			);
+		}
+
+		// Refuse to install a Helper Script that is too large.
+		$helper_script_size = strlen( $script_body );
+		if ( $helper_script_size > static::MAX_FILESIZE ) {
+			return new \WP_Error(
+				'too_big',
+				"Helper script is bigger ($helper_script_size bytes) " .
+				'than the max. size (' . static::MAX_FILESIZE . ' bytes)',
+				array( 'status' => 413 )
+			);
+		}
+
+		// Replace '[wp_path]' in the Helper Script with the WordPress installation location. Allows the Helper Script
+		// to find WordPress.
+		$wp_path_marker = '[wp_path]';
+		try {
+			$normalized_abspath = addslashes( Throw_On_Errors::t_realpath( ABSPATH ) );
```

Comment:
Thanks, possibly this could be improved later, but here I'm just preserving how it was all being done before, in an attempt to not break anything.

Changing up how that `[wp_path]` gets injected would involve us having to change the helper script itself, then have/support two versions of the helper while the plugin changes get gradually rolled out, etc.

I don't see this as a security issue either, as the user with NUL bytes in their `ABSPATH` just won't get their website backed up/restored. So, not quite ready to do all of this now.